![](../storage/banners/40_assumptions.png)

In [1]:
from __init__ import *

Инициализация проекта kiam-formation | Контекст: kiamformation


##### Быстродействие

`clip`: своё лучше для обычных чисел

In [43]:
import numpy as np
import time

n = int(1e6)

# Моя функция
def clip(a: float, bot: float, top: float) -> float:
    if a < bot:
        return bot
    if a > top:
        return top
    return a

t0 = time.time()
for _ in range(n):
    tmp = clip(-1, 0, 2)
print(f"my-clip: {time.time() - t0} секунд")

t0 = time.time()
for _ in range(n):
    tmp = np.clip(-1, 0, 2)
print(f"np.clip: {time.time() - t0} секунд")

my-clip: 0.14356112480163574 секунд
np.clip: 3.1530544757843018 секунд


##### Кватернионы в ПО

In [ ]:
# import numpy as np
# s0 = dir(np)

In [ ]:
# s = dir(quaternion.np)
# for i in s0:
#     s.remove(i)
# s

**Инициализация**

In [100]:
import numpy as np
import quaternion

for q in [np.quaternion(1, 0, 0, 0),
          np.quaternion(1, 2, 3, 4),
          np.quaternion(5, 0, 0),
          np.quaternion(5),
          np.quaternion()]:
    print(f"q = {q}")

q = np.quaternion(1, 2, 3, 4)

q = quaternion(1, 0, 0, 0)
q = quaternion(1, 2, 3, 4)
q = quaternion(0, 5, 0, 0)
q = quaternion(5, 0, 0, 0)
q = quaternion(0, 0, 0, 0)


**Преобразования типов**

In [66]:
a = quaternion.as_float_array(q)
b = quaternion.np.array(q)
c = quaternion.np.asarray(q)
d = q.components

print(f"Как numpy-массив: {a}")
print(type(a))
print(f"Как numpy-массив: {b}")
print(type(b))
print(f"Как numpy-массив: {c}")
print(type(c))
print(f"Как numpy-массив: {d}")
print(type(d))

Как numpy-массив: [1. 2. 3. 4.]
<class 'numpy.ndarray'>
Как numpy-массив: quaternion(1, 2, 3, 4)
<class 'numpy.ndarray'>
Как numpy-массив: quaternion(1, 2, 3, 4)
<class 'numpy.ndarray'>
Как numpy-массив: [1. 2. 3. 4.]
<class 'numpy.ndarray'>


**Преобразовани математические**

In [101]:
a = q.real
b = q.vec
w = q.w
x = q.x
y = q.y
z = q.z

print(f"Скалярная часть: {a}")
print(type(a))
print(f"Векторная часть: {b}")
print(type(b))
print(f"Составлющие: {w} {x} {y} {z}")
print(type(w), type(x), type(y), type(z))

Скалярная часть: 1.0
<class 'float'>
Векторная часть: [2. 3. 4.]
<class 'numpy.ndarray'>
Составлющие: 1.0 2.0 3.0 4.0
<class 'float'> <class 'float'> <class 'float'> <class 'float'>


In [10]:
a = quaternion.np.normalized(q)

print(f"Нормированный кватернион: {a}")
print(type(a))

Нормированный кватернион: quaternion(0.182574185835055, 0.365148371670111, 0.547722557505166, 0.730296743340221)
<class 'quaternion.quaternion'>


In [99]:
a = quaternion.np.abs(q)
b = quaternion.np.absolute(q)
d = q.abs()

c = quaternion.np.norm(q)

print(f"Модуль: {a} | {b} | {d}")
print(type(a), type(b), type(d), '\n')

print(f"Норма: {c} (√c = {np.sqrt(c)})")
print(type(c))

Модуль: 5.477225575051661 | 5.477225575051661 | 5.477225575051661
<class 'numpy.float64'> <class 'numpy.float64'> <class 'float'> 

Норма: 30.0 (√c = 5.477225575051661)
<class 'numpy.float64'>


In [27]:
a = quaternion.np.angle_of_rotor(q)
b = q.angle()

print(f"Угол поворота: {a} | {b}")
print(type(a), type(b))

Угол поворота: 4.389231563842282 | 4.389231563842282
<class 'numpy.float64'> <class 'float'>


In [96]:
a = q.conj()
b = q.conjugate()

с = q.inverse()
d = q.reciprocal()

print(f"Сопряженный: {a} | {b}")
print(type(a), type(b), '\n')

print(f"Обратный: {с} | {d}")
print(type(с), type(d))

Сопряженный: quaternion(1, -2, -3, -4) | quaternion(1, -2, -3, -4)
<class 'quaternion.quaternion'> <class 'quaternion.quaternion'> 

Обратный: quaternion(0.0333333333333333, -0.0666666666666667, -0.1, -0.133333333333333) | quaternion(0.0333333333333333, -0.0666666666666667, -0.1, -0.133333333333333)
<class 'quaternion.quaternion'> <class 'quaternion.quaternion'>


##### Проверка кватернионного умножения

###### Символьные выражения

In [3]:
from sympy import *

q1 = kf.get_vars("q^1", 4, numb=False)
q2 = kf.get_vars("q^2", 4, numb=False)
q1v = Matrix(q1[1:4])
q2v = Matrix(q2[1:4])

print("Скалярная часть:")
q1[0]*q2[0] - q1v.dot(q2v)

Скалярная часть:


q^1_0*q^2_0 - q^1_x*q^2_x - q^1_y*q^2_y - q^1_z*q^2_z

In [4]:
print("Векторная часть:")
q1v*q2[0] + q2v*q1[0] + q1v.cross(q2v)

Векторная часть:


Matrix([
[q^1_0*q^2_x + q^1_x*q^2_0 + q^1_y*q^2_z - q^1_z*q^2_y],
[q^1_0*q^2_y - q^1_x*q^2_z + q^1_y*q^2_0 + q^1_z*q^2_x],
[q^1_0*q^2_z + q^1_x*q^2_y - q^1_y*q^2_x + q^1_z*q^2_0]])

###### Старый код

In [46]:
import numpy as np

def vec2quat(v) -> list:
    if len(v) != 3:
        raise ValueError(f"Подаётся вектор длинны {len(v)}, требуется длинна 3!")
    if np.linalg.norm(v) > 1:
        return [0] + list(np.array(v)/np.linalg.norm(v))
    else:
        return [np.sqrt(1 - np.linalg.norm(v)**2), v[0], v[1], v[2]]

def q_dot(L1, L2):
    return np.array([L1[0] * L2[0] - L1[1] * L2[1] - L1[2] * L2[2] - L1[3] * L2[3],
                     L1[0] * L2[1] + L1[1] * L2[0] + L1[2] * L2[3] - L1[3] * L2[2],
                     L1[0] * L2[2] + L1[2] * L2[0] + L1[3] * L2[1] - L1[1] * L2[3],
                     L1[0] * L2[3] + L1[3] * L2[0] + L1[1] * L2[2] - L1[2] * L2[1]])

def quart2dcm(L) -> np.ndarray:
    w, x, y, z = L
    A = np.eye(3)
    A[0][0] = 1 - 2 * y ** 2 - 2 * z ** 2
    A[0][1] = 2 * x * y + 2 * z * w
    A[0][2] = 2 * x * z - 2 * y * w
    A[1][0] = 2 * x * y - 2 * z * w
    A[1][1] = 1 - 2 * x ** 2 - 2 * z ** 2
    A[1][2] = 2 * y * z + 2 * x * w
    A[2][0] = 2 * x * z + 2 * y * w
    A[2][1] = 2 * y * z - 2 * x * w
    A[2][2] = 1 - 2 * x ** 2 - 2 * y ** 2
    return A

def check_quat(q1, q2, q_dot, quart2dcm, vec2quat):
    print(f"λ₁ = {q1}")
    if isinstance(q1, list | np.ndarray):
        print(f"q3->q4: {vec2quat(q1[1:4])}")
        if isinstance(vec2quat(q1[1:4]), list | np.ndarray):
            print(f"модуль {vec2quat(q1[1:4])} = {np.linalg.norm(vec2quat(q1[1:4]))})")
    print(f"λ₂ = {q2}")
    print(f"λ₁∘λ₂ = {q_dot(q1, q2)}\n")

    A = quart2dcm(q1)
    print(f"A(λ₁) = \n{A} (type(A)={type(A)}) \nДетерминант {np.linalg.det(A)}, \nAA.T=\n{np.round(A @ A.T).astype(np.int64)}")

qs = [np.random.rand(4) for _ in range(2)]
for i in range(2):
    qs[i] /= np.linalg.norm(qs[i])

check_quat(qs[0], qs[1], q_dot, quart2dcm, vec2quat)

λ₁ = [0.50963435 0.55272453 0.57937229 0.31479545]
q3->q4: [0.5096343479799098, 0.5527245303128928, 0.5793722880928656, 0.31479545222179095]
модуль [0.5096343479799098, 0.5527245303128928, 0.5793722880928656, 0.31479545222179095] = 0.9999999999999999)
λ₂ = [0.57007807 0.63390766 0.5152193  0.08786999]
λ₁∘λ₂ = [-0.38600981  0.526878    0.74384424  0.14174536]

A(λ₁) = 
[[ 0.13046315  0.9613277  -0.2425457 ]
 [ 0.3196054   0.19079883  0.92814233]
 [ 0.93852637 -0.19860729 -0.28235331]] (type(A)=<class 'numpy.ndarray'>) 
Детерминант 1.0000000000000007, 
AA.T=
[[1 0 0]
 [0 1 0]
 [0 0 1]]


###### Новый код

In [47]:
import numpy as np
import quaternion

def vec2quat(v):
    """Перевод вектора кватерниона в кватернион"""
    if isinstance(v, np.ndarray | list | quaternion.quaternion):
        return np.quaternion(*v)
    else:
        from sympy import Matrix
        return Matrix([sqrt(1 - v.dot(v)), v[0], v[1], v[2]])

def q_dot(q1, q2):
    """Умножение кватернионов длины 4"""
    if isinstance(q1, quaternion.quaternion):
        return q1 * q2
    else:
        from sympy import Matrix
        q1v = Matrix(q1[1:4])
        q2v = Matrix(q2[1:4])
        scalar = q1[0]*q2[0] - q1v.dot(q2v)
        vector = q1v*q2[0] + q2v*q1[0] + q1v.cross(q2v)
        return Matrix([scalar, vector[0], vector[1], vector[2]])

def quart2dcm(q):
    """Матрицу поворота из кватерниона поворота"""
    if isinstance(q, quaternion.quaternion):
        return quaternion.as_rotation_matrix(q)
    else:
        from sympy import Matrix, sqrt
        w, x, y, z = q
        return Matrix([[1 - 2*y**2 - 2*z**2, 2*x*y + 2*z*w, 2*x*z - 2*y*w],
                       [2*x*y - 2*z*w, 1 - 2*x**2 - 2*z**2, 2*y*z + 2*x*w],
                       [2*x*z + 2*y*w, 2*y*z - 2*x*w, 1 - 2*x**2 - 2*y**2]])

for i in range(2):
    qs[i] = np.quaternion(*qs[i])
check_quat(qs[0], qs[1], q_dot, quart2dcm, vec2quat)

λ₁ = quaternion(0.50963434797991, 0.552724530312893, 0.579372288092866, 0.314795452221791)
λ₂ = quaternion(0.570078071786166, 0.633907663196747, 0.51521930422025, 0.0878699901434395)
λ₁∘λ₂ = quaternion(-0.386009809429407, 0.526877996461123, 0.743844241487517, 0.141745364210778)

A(λ₁) = 
[[ 0.13046315  0.3196054   0.93852637]
 [ 0.9613277   0.19079883 -0.19860729]
 [-0.2425457   0.92814233 -0.28235331]] (type(A)=<class 'numpy.ndarray'>) 
Детерминант 1.0000000000000007, 
AA.T=
[[1 0 0]
 [0 1 0]
 [0 0 1]]


λ₁ = [0.01206783 0.61613075 0.63217471 0.46967265]
q3->q4: [0.012067828774893265, 0.6161307531238012, 0.6321747111229158, 0.4696726489592675]
модуль [0.012067828774893265, 0.6161307531238012, 0.6321747111229158, 0.4696726489592675] = 1.0)
λ₂ = [0.37735788 0.7657617  0.50802446 0.11454778]
λ₁∘λ₂ = [-0.84221562  0.07555189  0.53376778  0.00753134]

A(λ₁) = 
[[-0.24047453  0.79034042  0.56350157]
 [ 0.7676687  -0.200419    0.60870106]
 [ 0.59401748  0.57895962 -0.55852394]] 
Детерминант 1.000000000000001, 
AA.T=
[[1 0 0]
 [0 1 0]
 [0 0 1]]

In [73]:
a

quaternion(0, 0, 0.5, 0)

In [75]:
g = quaternion.as_float_array(a)
g[[1, 2, 3]]

array([0. , 0.5, 0. ])

In [ ]:
dir(a)

In [69]:
type(a.x)

float

In [80]:
from sympy import *

m = Matrix([1, 2, 3, 4, 5])
type(m[1:3])

list

In [ ]:
import sympy

dir(sympy)

### a * 2

In [ ]:
dir(m)